File: movimentacoes.py
"""Histórico de movimentações — usado pelo ERP e espelhado no app (movimentacoes.js).

Registrar o que aconteceu é barato e responde sozinho as perguntas que
aparecem depois ("quem cancelou essa visita?", "quando isso virou Em
Andamento?"). Nada aqui pode derrubar a operação: se o registro falhar, a
ação que o usuário fez continua valendo.
"""
import streamlit as st

ROTULOS = {
    "criou": "criou",
    "status": "mudou o status",
    "comentario": "comentou",
    "midia": "anexou mídia",
    "editou": "editou",
    "concluiu": "concluiu",
}


def registrar(supabase, tipo, referencia_id, acao, *, usuario=None, de=None, para=None, detalhe=None):
    try:
        supabase.table("movimentacoes").insert({
            "tipo": tipo, "referencia_id": int(referencia_id),
            "usuario": usuario or st.session_state.get("usuario_logado", "Sistema"),
            "acao": acao, "de": de, "para": para, "detalhe": detalhe,
        }).execute()
    except Exception:
        pass  # histórico é registro, não pode travar o que o usuário fez


def listar(supabase, tipo, referencia_id, limite=40):
    try:
        return supabase.table("movimentacoes").select("*").eq("tipo", tipo).eq(
            "referencia_id", int(referencia_id)).order("criado_em", desc=True).limit(limite).execute().data or []
    except Exception:
        return []


def frase(mov):
    """Uma linha legível a partir do registro."""
    quem = mov.get("usuario") or "Alguém"
    acao = ROTULOS.get(mov.get("acao"), mov.get("acao") or "mexeu")
    if mov.get("acao") == "status" and mov.get("para"):
        de = f" (era {mov['de']})" if mov.get("de") else ""
        return f"{quem} {acao} para **{mov['para']}**{de}"
    if mov.get("detalhe"):
        return f"{quem} {acao}: {mov['detalhe']}"
    return f"{quem} {acao}"


def render(supabase, tipo, referencia_id, titulo="🕓 Histórico de movimentações"):
    """Bloco de histórico pronto pra colar em qualquer tela do ERP."""
    import pandas as pd
    movs = listar(supabase, tipo, referencia_id)
    with st.expander(f"{titulo} ({len(movs)})", expanded=False):
        if not movs:
            st.caption("Nada registrado ainda.")
            return
        for m in movs:
            quando = pd.to_datetime(m.get("criado_em"), errors="coerce")
            carimbo = quando.strftime("%d/%m/%Y %H:%M") if pd.notna(quando) else ""
            st.markdown(f"- {frase(m)}  <span style='color:#888;font-size:0.85em;'>· {carimbo}</span>",
                        unsafe_allow_html=True)

File: docs/js/movimentacoes.js
import { getSupabase } from "./supabase-client.js";

// Espelho do movimentacoes.py do ERP: os dois escrevem na mesma tabela, então
// o histórico mostra o que o Breno fez no admin e o que o instalador fez no
// app, na mesma linha do tempo.
const TABELA = "movimentacoes";

const ROTULOS = {
  criou: "criou",
  status: "mudou o status",
  comentario: "comentou",
  midia: "anexou mídia",
  editou: "editou",
  concluiu: "concluiu",
  confirmou: "confirmou presença",
  remarcar: "pediu para remarcar",
};

/** Registra a movimentação. Nunca lança: histórico é registro, não pode
 *  impedir o instalador de trabalhar (ainda mais offline, sem sinal). */
export async function registrar(tipo, referenciaId, acao, { usuario, de, para, detalhe } = {}) {
  try {
    const supabase = getSupabase();
    await supabase.from(TABELA).insert({
      tipo, referencia_id: referenciaId, usuario: usuario || null,
      acao, de: de || null, para: para || null, detalhe: detalhe || null,
    });
  } catch (e) {
    /* sem sinal: segue o jogo */
  }
}

export async function listar(tipo, referenciaId, limite = 30) {
  try {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from(TABELA).select("*")
      .eq("tipo", tipo).eq("referencia_id", referenciaId)
      .order("criado_em", { ascending: false }).limit(limite);
    if (error) throw error;
    return data || [];
  } catch (e) {
    return [];
  }
}

export function frase(m) {
  const quem = m.usuario || "Alguém";
  const acao = ROTULOS[m.acao] || m.acao || "mexeu";
  if (m.acao === "status" && m.para) {
    return `${quem} ${acao} para ${m.para}${m.de ? ` (era ${m.de})` : ""}`;
  }
  return m.detalhe ? `${quem} ${acao}: ${m.detalhe}` : `${quem} ${acao}`;
}

export function quando(m) {
  const d = new Date(m.criado_em);
  if (isNaN(d)) return "";
  return d.toLocaleString("pt-BR", { day: "2-digit", month: "2-digit", hour: "2-digit", minute: "2-digit" });
}
